File: frameworks/web-framework-cpp/web_framework_benchmark/executors/src/JsonExecutor.cpp
#include "JsonExecutor.hpp"

#include <charconv>

namespace executor
{
	void JsonExecutor::init(const framework::utility::ExecutorSettings& settings)
	{
		dataset = framework::JsonParser(settings.getFile("dataset.json")).getParsedData(true);
		items = dataset.get<std::vector<framework::JsonObject>>();
	}

	void JsonExecutor::doGet(framework::HttpRequest& request, framework::HttpResponse& response)
	{
		const std::string& queryParameter = request.getQueryParameters().at("m");
		int count = request.getRouteParameter<int>("count");
		int multiplier;
		framework::JsonBuilder result;
		std::vector<framework::JsonObject> resultItems;

		resultItems.reserve(count);

		std::from_chars(queryParameter.data(), queryParameter.data() + queryParameter.size(), multiplier);

		for (int i = 0; i < count; i++)
		{
			framework::JsonObject temp(items[i]);

			temp["total"] = temp["price"].get<int>() * temp["quantity"].get<int>() * multiplier;

			resultItems.emplace_back(std::move(temp));
		}

		result["count"] = count;
		result["items"] = std::move(resultItems);

		response.setBody(result);
	}

	DEFINE_EXECUTOR(JsonExecutor);
}

File: frameworks/web-framework-csharp/web_framework_benchmark/Executors/JsonExecutor.cs
namespace Executors;

using Framework;
using System.Text.Json.Nodes;

public class JsonExecutor : StatelessExecutor
{
	private JsonArray? _items;

	public override void Init(ExecutorSettings settings) => _items = JsonNode.Parse(settings.GetFile("dataset.json"))!.AsArray();

	public override void DoGet(HttpRequest request, HttpResponse response)
	{
		int multiplier = int.Parse(request.GetQueryParameters()["m"]);
		int count = request.GetRouteParameter<int>("count");
		var result = new
		{
			count,
			items = new JsonArray()
		};

		for (int i = 0; i < count; i++)
		{
			JsonObject temp = _items!.ElementAt(i)!.AsObject().DeepClone().AsObject();

			temp["total"] = temp["price"]!.GetValue<int>() * temp["quantity"]!.GetValue<int>() * multiplier;

			result.items.Add(temp);
		}

		response.SetBody(result);
	}
}
